### ISSUE TITLE
Array-chunks Iterator API to cKDTree.query_pairs

### ISSUE THREAD

**Issue by rainwoodman** [scipy/scipy#5710]
The current version of cKDTree.query_pairs contains an output argument, where 'set' and 'array' outputs are supported. The array output format was intented for larger number of pairs. However when the number of pairs exceed the memory bound, the array interface becomes insufficient as well. 
An example would be enumerating pairs of galaxies (millions) for calculating two point and three point functions at large separations.

Another limitation is that the current query_pairs API does not allow a dual tree iteration.

I propose we modify query_pairs, allowing the following use case:
```
for pairs in tree.query_pairs(r, p=2, eps=0, other=other, output_type='chunk', chunksize=100000):
     assert pairs.shape[1] == 2
     assert pairs.shape[0] <= 100000
```

The additional API likely requires some non-trivial modifications to the current query_pair implementation to allow efficient chunking.

**Comment by rainwoodman** [comment]
@sturlamolden : what do you think?

**Comment by sturlamolden** [comment]
I don't really have an opinion on this, so I'll just be neutral.

**Comment by gatagat** [comment]
Any kind of iterator/chunking would be a great improvement. The current behavior causes troubles especially when processing datasets where you do not know the number of points and their distances beforehand - you do not know how many pairs you will get but if your try to get them it might not ever return.

**Comment by rainwoodman** [comment]
Last time I played with this, the biggest implementation challenge is to be able to save the state of the dual-tree walk for a later continuation at C level. 

I think we will likely need to unroll the current recursive tree walk into a FIFO queue implementation.

**Comment by sturlamolden** [comment]
Using two threads would also solve this. Instead of saving the state or unrolling the recursion, spawn one thread to walk the trees and let the main thread deliver the data to the user via an interator. A simple design would be to use a ringbuffer of, say, 16 data chunks, each containing up to a maximum of 1024 pairs. If the main thread (iterator) does not deliver the data fast enough, the tree walker just goes to sleep until a new buffer is ready on the tail of the queue. Sleeping on the queue will effectively save the state of the tree walker.

**Comment by rainwoodman** [comment]
Indeed. I thought about using a thread -- I used threads for the chunk interface in my own version of kdtree. I nevertheless felt uncomfortable invoking threads for such a low-level, computational intensive operation though.
An additional benefit of unrolling the recursion is some potential performance gain.

**Comment by sturlamolden** [comment]
Since we are on C++, there is also a more crazy solution: Boost yield-type coroutine 😀
http://aldrin.co/coroutine-basics.html
http://www.boost.org/doc/libs/1_63_0/libs/coroutine/doc/html/index.html

**Comment by sturlamolden** [comment]
For those unfamiliar with Boost.Coroutine and C++, it is basically the same as a Python generator implemented with a yield statement.

**Comment by sturlamolden** [comment]
Switch performance (the interesting part for computational tasks) seems rather efficient. 30 to 50 clock cycles is not bad. With some chunking this can be amortized to nothing.

http://www.boost.org/doc/libs/1_63_0/libs/coroutine/doc/html/coroutine/performance.html

**Comment by rainwoodman** [comment]
It sounds like the state of the python interpreter's stack is stored in the
context as well. But we cannot always assume the interpreter state is
consistent between switches.

For example, if we obtain the generator, iterate for a few steps, save it
to a module level variable, then return, return, ... return, later during
another module's initialization we continue the iteration. The stack saved
the context will surely be referring to non-existing variables.

I think we can very likely prove we never access those expired pointers.
But in my perspective this doesn't smell very nice.
